L2 rework

tbursztyka opened the pull request:
Changing the way packet are sent: instead of going like net_if_send_data -> l2 send -> net_if tx queue -> net_if_tx -> driver send
it will be now:
net_if_send _data -> net_if tx queue -> net_if_tx ->  l2 send ->  driver send

This will enable removing ll reserve from l2 as well as doing interesting optimization in L2 side.

- Now L2 is responsible for unrefing the packet on success. (net_if still keeps that responsibility in case of error) so drivers do not have to care about it.

- l2 send does not return a verdict, but a int: 0+ (lenght of sent packet) on succes, negative errno otherwise.

- driver API have to provide their own send/tx function, relatively to their respective L2. There is no more the net_if_api send function. That will permit l2 to have whatever signature they need for that function (15.4 has done so for quite a while, but it was thus a hack).

@tgorochowik Could you check if the sam gmac driver still works fine? It's a bit different in there, as it queues tx pkt. I hope I did not break it.
@Vudentz It was relatively simple in bt L2 to do the change, but I could not test it. I believe it's fine though.


All of this is part of the preliminary work to revise net_pkt buffer management totally.

Commit:
net/ethernet: Make ethernet allocating its own frag for the header

This is currently unoptimized, as all frags are allocated with relevant
ll reserve for such header space. However, this is the first step
towards getting rid of that ll reserve concept everywhere.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ethernet: Centralize where and when header is filled in

Current code generating Ethernet header is scattered all over the place,
sometimes in functions that are supposed to check something (and not
filling the header). Not to say about innefficiency.

Src ll address does not need to be set in L2 as net_if.c handles that
already.

Broadcast dst ll address is the same in ipv4 or ipv6, thus factorizing.
In each case, multicast is filled in only at the relevant place.

This is the first step towards changing L2 sending logic, when L2 send
API function will be the only point of sending. The redirection from
driver to L2 again (which finally uses the right device API function to
send) it a temporary hack.

This simplifies the code but will also enable using statically
allocated net_buf and ethernet header payload buffer afterwards.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/arp: ARP packet can be queued right away

Now that sending is done at last time, in one pass, no need to go
through net_if_send_data here.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ethernet: Move sending to a one-pass logic

Currently, first part is done in L2's send, then the next one in
ethernet device driver net_if send function. That last one was already
moved to a L2 based implementation. Let's just move forward and place
the whole logic of the L2's send in that second function.

This is the first step, ethernet centric only, to move towards a
one-pass sending logic in net stack. In future, net_if's send will
disappear.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/iface: Switch fully to a one-pass sending logic in net_if

Now instead of such path:

net_if_send_data -> L2's send -> net_if tx_queue -> net_if_tx -> driver
net_if's send

It will be:

net_if_send_data -> net_if tx_queue -> net_if_tx -> L2's send -> driver
net_if's send

Only Ethernet is adapted, but 15.4 and bt will follow up.
All Ethernet drivers are made compatible with that new scheme also.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Switch to L2 sending path

As for Ethernet, up to ieee802154 L2's send to actually sent the packet.
It's currently unoptimized as 6lo compression, 15.4 fragmentation and so
on will reallocate net_buf etc... but it's the first step towards
removing ll reserve space and more.

Applying changes to Openthread L2 as well.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/bluetooth: Switch to L2 sending path

This is a much more trivial move than in Ethernet or ieee802154 L2s.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/dummy: Switch to L2 sending path

And adapt loopback and slip drivers relevantly

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/net: Apply L2 change everywhere where relevant

Many tests use either Ethernet or Dummy L2 and as such require
modifications towards the driver API on their fake devices.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/iface: Finally remove the send function from net_if_api

And apply that to modem driver setting that pointer to NULL.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tgorochowik reviewed (dismissed):
There are some left-over references to `api->send`, and the code doesn't compile:
```
In file included from /zephyr-path/subsys/net/ip/net_if.c:15:
/zephyr-path/subsys/net/ip/net_if.c: In function ‘init_iface’:
/zephyr-path/subsys/net/ip/net_if.c:246:17: error: ‘const struct net_if_api’ has no member named ‘send’
   NET_ASSERT(api->send);
                 ^~
/zephyr-path/include/net/net_core.h:64:9: note: in definition of macro ‘NET_ASSERT’
   if (!(cond)) {          \
         ^~~~
make[2]: *** [zephyr/subsys/net/ip/CMakeFiles/subsys__net__ip.dir/build.make:89: zephyr/subsys/net/ip/CMakeFiles/subsys__net__ip.dir/net_if.c.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:1808: zephyr/subsys/net/ip/CMakeFiles/subsys__net__ip.dir/all] Error 2
```

Therefore I am unable to test it on sam gmac.

tbursztyka commented:
It always amaze me how local sanitycheck running never finds out such obvious thing...

rveerama1 commented:
Overall looks good.

jukkar commented on include/net/ethernet.h:
```diff
@@ -329,7 +329,6 @@ struct ethernet_context {
  */
```
The "net/ethernet: Remove as many ifdef for CONFIG_NET_VLAN as possible" could be sent to upstream separately.

tgorochowik commented:
I just tested this on sam e70 with the `dhcpv4_client` sample with `NET_BUF_POOL_USAGE` and `NET_SHELL` enabled.

It is quite weird, your changes should only affect the the TX side in that driver, but zephyr seems to be leaking RX buffers.

After pinging zephyr with `ping 192.168.0.101 -i 0.00001` for a few minutes, the buffer pools look like this:
```
uart:~$ net mem
Fragment length 128 bytes
Network buffer pools:
Address         Total   Avail   Name
0x20407638      14      40      RX
0x20407654      14      21      TX
0x204076e4      36      1       RX DATA (rx_bufs)
0x2040770c      36      36      TX DATA (tx_bufs)
```

And it stops responding to further pings.

I am going to dismiss my review for now, as this might be a broader issue. The TX side that you changed here seems to work fine.

tbursztyka commented:
@tgorochowik thanks for retesting this. Looks like I may have messed up your driver because icmp echo replies are just built on rx echo requests packets, so these net_pkt do come for RX reserve. I'll check

pfalcon commented:
> Looks like I may have messed up your driver because icmp echo replies are just built on rx echo requests packets, so these net_pkt do come for RX reserve. 

Well, that's bad. That's a hack, and we won't have a clean/reliable stack until we give that "optimization" up. RX packets are for RX, TX for TX. They may come from *absolutely* different pool implementations. E.g., from hardware-backed pool, allowing only RX or only TX, respectively.

tbursztyka commented:
@pfalcon 

> Well, that's bad. That's a hack, and we won't have a clean/reliable stack until we give that "optimization" up. RX packets are for RX, TX for TX. They may come from absolutely different pool implementations. E.g., from hardware-backed pool, allowing only RX or only TX, respectively.

That has been like that almost since day 1. Actually, this RX/TX differentiation is going to disappear mostly in next to come buffer management rework (using various pools will remain). TX or RX you don't care much: if one of the pool if fully used, the system will eventually be unresponsive (normally, if it's not due to a leak - unlike what I did here - but an overload, timeout will kick in and system will start to response again). So it appears like a hack because of this tx/rx differentiation, but it's actually not a hack.

pfalcon commented:
> That has been like that almost since day 1.

Yeah, I know.

>  TX or RX you don't care much: if one of the pool if fully used, the system will eventually be unresponsive

Well, user may not care, but hardware definitely does. The processes behind RX and TX are different, and buffer management is part of that process. And there's no need to "help" system become unresponsive by misusing buffers. What should be done with RX buffers is quickly parse/copy data out of it, and get buffer back into the pool to accept new data. Holding onto the RX buffer for longer than needed, misusing it to stuff TX data and putting it into TX queue (which can be arbitrarily long) should be big no-no.

Again, IMHO we won't have a decent stack until we accept things above and implement it like that.


Otherwise, this is "oh, BTW" note for now, it just your comment triggered things I had in mind for a while.

tbursztyka commented:
> Well, user may not care, but hardware definitely does.

Not really. Hardware buffers, or local driver buffers are always unrelated to net_pkt buffers. We always copy (either on tx or rx path, in the drivers). On rx path, if drivers are not able to get a free net_pkt or the necessary net_buf , it will have to drop the frame from its internal buffers. That's fully nominal.
This TX/RX pool split really is misleading, hopefully we'll get rid of that.

pfalcon commented:
> Hardware buffers, or local driver buffers are always unrelated to net_pkt buffers. We always copy

Ok, so let me reformulate: all the above true if we want [one sweet day] to achieve zero-copy networking. Given that we don't have a reliable networking so far, I'm happy with anything which helps to achieve that first.

galak commented:
@dleach02 I know you had some thoughts on the L2 layering, can you take a look and comment.

tbursztyka commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -436,9 +436,9 @@ static bool eth_get_ptp_data(struct net_if *iface, struct net_pkt *pkt,
 }
```
It was the driver I used to test each steps. I could try moving these changes to the patch that applies changes to all other ethernet drivers.

jukkar replied:
> I could try moving these changes to the patch that applies changes to all other ethernet drivers.

That would indeed make the patches a bit more clearer.

tbursztyka commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -911,14 +911,13 @@ static struct device *eth_mcux_get_ptp_clock(struct device *dev)
-	.get_capabilities = eth_mcux_get_capabilities,
-
+	.iface_api.init		= eth_iface_init,
+	.iface_api.send		= net_eth_send,
```
iface_api.send is never set by L2. Well here anyway it's temporary change since in the end that send pointer disappears altogether

jukkar replied:
> iface_api.send is never set by L2. Well here anyway it's temporary change since in the end that send pointer disappears altogether

As the value is set to be the generic L2 send function, I was wondering if the L2 to set it automatically instead of each driver setting it. But if the field is going away, then just skip this question.

tbursztyka commented:
@tgorochowik I retried your test, and I never get what you have. However what I get at some point is the impossibility to reach the target. And from the uart console I can see it cannot go out either. Like if the ethernet device would be dead. And the same behaviour happens on master too, so I don't think I introduce any regression myself with that PR.

jukkar commented:
> Like if the ethernet device would be dead.

I am seeing similar things in sam-e70 too. Ping fails and machine does not seem to be able to receive any packets

dleach02 commented:
@galak, one issue I have with the network stack (from 15.4 perspective) is that there is knowledge that seems to go the wrong direction. For example, the 802.15.4 drivers are already aware of what is binding with it via the NET_DEVICE_INIT() macro which in turn required that we modify all of the drivers to support either OpenThread or native 15.4. Ideally, the 15.4 drivers would not know (or care) what is binding above and instead just provide the interface API struct to the binding stack with some sort of opaque blob being supplied down to the driver as needed and have the bindings drive down instead of up. 

The other area that I haven't been able to dig into yet is that it was reported to me that it is difficult to insert proprietary L2 protocols into our system. This may be the same problem as referenced above but the reporter was likely trying to put a protocol on top of our 15.4 stack.

Commit:
net/pkt: Add a gptp marker to know whether pkt is a gptp message or not

Like ARP, GPTP works on top of Ethernet but is considered as a layer 2
protocol, so we cannot set an AF_* family type. Instead let's have a bit
telling that current packet is a gptp message (family has to be
AF_UNSPEC).

The bit is unionized with a TCP related bit: TCP cannot be present on an
AF_UNSPEC packet so there will not be any collision.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/gptp: Let Ethernet setting up the header

Instead of redoing what Ethernet L2 already does, let just create the
gptp message without any Ethenet header. Which one will be done as
sending phase by Ethernet L2 relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@dleach02 Ok, so nothing to be done in that PR. (here the point is just to change when a packet is being sent: from L2 instead of net_if as it is currently). But feel free to create an issue for 15.4 with your requirements, we could think about something later on that.

jukkar commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -492,6 +492,8 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
 	} else if (IS_ENABLED(CONFIG_NET_IPV6) &&
```
As part of the #11229 I introduced L2 protocol field that stores the IEEE 802.3 protocol type field that is found in Ethernet header, it is found in commit ac56085fc077298fdda6eb20f0a008375dcb9edf. We could use that here which would allow to avoid having this kind of gptp specific info. Of course the network driver would need to set this after receiving the packet, and corresponding protocol handler would need to set it in TX. As a short term solution doing it like this is ok too.

tbursztyka replied:
Myes, I have reviewed yet your RFC, but adding more stuff to net_pkt again is maybe not what we want. Also I am really not a fan of drivers that has to do an "L2 job" there (parsing eth header). But let's see the details in that PR.

jukkar replied:
If we introduce this "l2_proto" flag into net_pkt, we could remove separate setting for the protocol family as that could be deducted from the l2_proto field. But we can discuss that later in #11229

dleach02 commented:
@jukkar, I understand what the macro is doing. My only point is that it seems a bit reverse in that it exists in the driver and seems to bind up instead of binding down. As a result, the drivers have a bit too much information on what is binding to it and when we add new things on top like the hackish OpenThread, it complicates the drivers because you have to touch every driver just to add OpenThread. It may be that is the only place this where happen since other 15.4 facilitated protocols should bind in on top of our 15.4 stack...

jukkar commented:
There is a reference counting issue somewhere, the ref counts go to -1 in packet sending. This happens when I tried it in sam_e70_xplained and gptp sample app.

I generated the net_pkt debug log with this config
```
CONFIG_NET_PKT_LOG_LEVEL_DBG=y
CONFIG_NET_BUF_POOL_USAGE=y
CONFIG_LOG_INPLACE_PROCESS=y
```

Below is small snippet when the machine boots
```


[00:00:m00.0:00,4[m62] <inf> i2c_sam_twihs.i2c_sam_twihs_initialize: Device I2C_0 initialized
[[00:00:00.008,925] <dbg> infnet_pkt.net_pkt_init: (0x20402ie74): Allocating 32 RX (3584 bytes), [32 TX (35:84 bytes), 
[00:00:00.:019,[m435] <dbg> net_pkt.net_pkt_init: (0x20402e74):   32 RX data (0 bytes) and 32 TX data (0 bytes) buffers
[00:00:00.031,315] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [31] frag 0x20407300 reserve 0
[00:00:00.042,161] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.052,191] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [30] frag 0x20407318 reserve 0
[00:00:00.063,156] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.073,187] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [29] frag 0x20407330 reserve 0
[00:00:00.084,152] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.094,182] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [28] frag 0x20407348 reserve 0
[00:00:00.105,146] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.115,177] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [27] frag 0x20407360 reserve 0
[00:00:00.126,143] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.136,173] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [26] frag 0x20407378 reserve 0
[00:00:00.147,136] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.157,168] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [25] frag 0x20407390 reserve 0
[00:00:00.168,132] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.178,164] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [24] frag 0x204073a8 reserve 0
[00:00:00.189,127] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.199,159] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [23] frag 0x204073c0 reserve 0
[00:00:00.210,123] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.220,155] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [22] frag 0x204073d8 reserve 0
[00:00:00.231,118] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.241,150] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [21] frag 0x204073f0 reserve 0
[00:00:00.252,114] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.262,144] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [20] frag 0x20407408 reserve 0
[00:00:00.273,107] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.283,139] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [19] frag 0x20407420 reserve 0
[00:00:00.294,103] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.304,135] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [18] frag 0x20407438 reserve 0
[00:00:00.315,098] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.325,130] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [17] frag 0x20407450 reserve 0
[00:00:00.336,094] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.346,125] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [16] frag 0x20407468 reserve 0
[00:00:00.357,089] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.367,121] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [15] frag 0x20407480 reserve 0
[00:00:00.378,085] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:00.388,116] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): RDATA (rx_bufs) [14] frag 0x20407498 reserve 0
[00:00:00.399,080] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (rx_descriptors_init():277)
[00:00:03.220,399] <dbg> net_pkt.net_pkt_get_reserve_debug: (0x20402e74): TX [31] pkt 0x20407290 reserve 14 ref 1 (send_mldv2():175)
[00:00:03.231,864] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): TDATA (tx_bufs) [31] frag 0x20408600 reserve 14
[00:00:03.242,911] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (net_pkt_append():1198)
[00:00:03.252,588] <dbg> net_pkt.net_pkt_frag_add_debug: (0x20402e74): pkt 0x20407290 frag 0x20408600 (net_pkt_append:1203)
[00:00:03.263,424] <dbg> net_pkt.net_pkt_get_reserve_debug: (0x20402e74): TX [30] pkt 0x20407220 reserve 14 ref 1 (send_mldv2_raw():92)
[00:00:03.275,241] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): TDATA (tx_bufs) [30] frag 0x20408618 reserve 14
[00:00:03.286,285] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (net_ipv6_create():69)
[00:00:03.295,874] <dbg> net_pkt.net_pkt_frag_insert_debug: (0x20402e74): pkt 0x20407220 frag 0x20408618 (net_ipv6_create:74)
[00:00:03.306,889] <dbg> net_pkt.net_pkt_frag_add_debug: (0x20402e74): pkt 0x20407220 frag 0x20408600 (send_mldv2_raw:131)
[00:00:03.317,558] <dbg> net_pkt.net_pkt_compact: (0x20402e74): Compacting data in pkt 0x20407220
[00:00:03.326,118] <dbg> net_pkt.net_pkt_frag_del_debug: (0x20402e74): pkt 0x20407220 parent 0x20408618 frag 0x20408600 ref 1 (net_pkt_compact:1157)
[00:00:03.339,161] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408600 reserve 14
[00:00:03.350,134] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80):   ref 1 (ethernet_fill_header():408)
[00:00:03.360,248] <dbg> net_pkt.net_pkt_frag_insert_debug: (0x20401c80): pkt 0x20407220 frag 0x20408600 (ethernet_fill_header:462)
[00:00:03.371,768] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TX [30] pkt 0x20407220 ref 0 frags 0x20408600 (tx_completed():554)
[00:00:03.373,488] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408600
[00:00:03.372,642] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref 0 frags 0x20408618 (tx_completed():554)
[00:00:03.372,586] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618
[00:00:03.371,737] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref 0 frags 0x00000000 (tx_completed():554)
[00:00:03.371,690] <dbg> net_pkt.net_pkt_frag_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408600 ref 0 (net_pkt_unref_debug():823)
[00:00:03.384,264] <dbg> net_pkt.net_pkt_get_reserve_debug: (0x20401c80): RX [31] pkt 0x20406490 reserve 0 ref 1 (frame_get():1121)
[00:00:03.385,708] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80): RDATA (rx_bufs) [13] frag 0x204074b0 reserve 0
[00:00:03.386,666] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80):   ref 1 (frame_get():1163)
[00:00:03.385,910] <dbg> net_pkt.net_pkt_frag_insert_debug: (0x20401c80): pkt 0x20406490 frag 0x20407300 (frame_get:1171)
[00:00:03.396,557] <dbg> net_pkt.net_pkt_ref_debug: (0x20401c80): TX [30] pkt 0x20407220 ref 2 (eth_tx():1417)
[00:00:03.406,219] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TX [31] pkt 0x20407220 ref 0 frags 0x20408600 (ethernet_send():536)
[00:00:03.418,055] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [32] frag 0x20408600
[00:00:03.427,209] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref -1 frags 0x00000000 (ethernet_send():536)
[00:00:03.437,347] [1;31m<err> net_pkt.net_pkt_unref_debug: *** ERROR *** frag 0x20408600 is freed already (ethernet_send():536)[0m
[00:00:03.449,053] <dbg> net_pkt.net_pkt_frag_unref_debug: (0x20401c80): TDATA (tx_bufs) [32] frag 0x20408600 ref -1 (net_pkt_unref_debug():823)
[00:00:03.461,623] <dbg> net_pkt.net_pkt_print: (0x20401af4): TX 32 RX 31 RDATA 13 TDATA 32
[9[00:00:03.469,861] <dbg> n;et_pkt.net_pkt_unref_debug: (f0x20402e74): TX [32] pkt g0x2204407290 ref 0 frags 0x0000000e0 (t_send_mldv2():192)[
3[00:r00:$03.m482,444[1] <dbagt> n emt_pkt.net_pkt_get_reserve_debug: (0x20402e74): TX [32] pkt 0x20407290 reserve 14 ref 1 (send_mldv2():175)
[00:00:03.494,959] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): TDATA (tx_bufs) [31] frag 0x20408618 reserve 14
[00:00:03.506,006] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (net_pkt_append():1198)
[00:00:03.515,683] <dbg> net_pkt.net_pkt_frag_add_debug: (0x20402e74): pkt 0x20407290 frag 0x20408618 (net_pkt_append:1203)
[00:00:03.526,517] <dbg> net_pkt.net_pkt_get_reserve_debug: (0x20402e74): TX [31] pkt 0x20407220 reserve 14 ref 1 (send_mldv2_raw():92)
[00:00:03.538,336] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): TDATA (tx_bufs) [30] frag 0x20408600 reserve 14
[00:00:03.549,379] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (net_ipv6_create():69)
[00:00:03.558,967] <dbg> net_pkt.net_pkt_frag_insert_debug: (0x20402e74): pkt 0x20407220 frag 0x20408600 (net_ipv6_create:74)
[00:00:03.569,984] <dbg> net_pkt.net_pkt_frag_add_debug: (0x20402e74): pkt 0x20407220 frag 0x20408618 (send_mldv2_raw:131)
[00:00:03.580,651] <dbg> net_pkt.net_pkt_compact: (0x20402e74): Compacting data in pkt 0x20407220
[00:00:03.589,213] <dbg> net_pkt.net_pkt_frag_del_debug: (0x20402e74): pkt 0x20407220 parent 0x20408600 frag 0x20408618 ref 1 (net_pkt_compact:1157)
[00:00:03.602,255] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618 reserve 14
[00:00:03.613,229] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80):   ref 1 (ethernet_fill_header():408)
[00:00:03.623,343] <dbg> net_pkt.net_pkt_frag_insert_debug: (0x20401c80): pkt 0x20407220 frag 0x20408618 (ethernet_fill_header:462)
[00:00:03.634,863] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TX [31] pkt 0x20407220 ref 0 frags 0x20408618 (tx_completed():554)
[00:00:03.636,583] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618
[00:00:03.635,737] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref 0 frags 0x20408600 (tx_completed():554)
[00:00:03.635,682] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408600
[00:00:03.634,832] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref 0 frags 0x00000000 (tx_completed():554)
[00:00:03.634,784] <dbg> net_pkt.net_pkt_frag_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618 ref 0 (net_pkt_unref_debug():823)
[00:00:03.647,344] <dbg> net_pkt.net_pkt_ref_debug: (0x20401c80): TX [31] pkt 0x20407220 ref 2 (eth_tx():1417)
[00:00:03.656,978] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TX [32] pkt 0x20407220 ref 0 frags 0x20408618 (ethernet_send():536)
[00:00:03.668,814] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [32] frag 0x20408618
[00:00:03.677,967] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref -1 frags 0x00000000 (ethernet_send():536)
[00:00:03.688,082] [1;31m<err> net_pkt.net_pkt_unref_debug: *** ERROR *** frag 0x20408618 is freed already by tx_completed():554 (ethernet_send():536)[0m
[00:00:03.701,711] <dbg> net_pkt.net_pkt_frag_unref_debug: (0x20401c80): TDATA (tx_bufs) [32] frag 0x20408618 ref -1 (net_pkt_unref_debug():823)
[9[[00:0:0:003.7814,4286] <;d1bg> rnet_pkt.net_pkt_unref_d_ebug: (0x20402e74): TX [33] 2pk0t6 0x20407290 rleefd 0 f rx_comags 0x00000000 (send_mldv2():192))
0[00:[00:03.728,182] <dbg> net_pkt.net_pkt_get_reserve_debug: (0x20402e74): TX [33] pkt 0x20407290 reserve 14 ref 1 (net_ipv6_send_ns():1962)
[00:00:03.740,448] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): TDATA (tx_bufs) [31] frag 0x20408600 reserve 14
[00:00:03.751,495] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (net_ipv6_send_ns():1967)
[00:00:03.761,344] <dbg> net_pkt.net_pkt_frag_add_debug: (0x20402e74): pkt 0x20407290 frag 0x20408600 (net_ipv6_send_ns:1973)
[00:00:03.772,412] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618 reserve 14
[00:00:03.783,376] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80):   ref 1 (ethernet_fill_header():408)
[00:00:03.793,489] <dbg> net_pkt.net_pkt_frag_insert_debug: (0x20401c80): pkt 0x20407290 frag 0x20408618 (ethernet_fill_header:462)
[00:00:03.805,008] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TX [33] pkt 0x20407290 ref 0 frags 0x20408618 (tx_completed():554)
[00:00:03.806,731] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618
[00:00:03.805,885] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref 0 frags 0x20408600 (tx_completed():554)
[00:00:03.805,828] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408600
[00:00:03.804,979] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref 0 frags 0x00000000 (tx_completed():554)
[00:00:03.804,929] <dbg> net_pkt.net_pkt_frag_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618 ref 0 (net_pkt_unref_debug():823)
[00:00:03.817,494] <dbg> net_pkt.net_pkt_ref_debug: (0x20401c80): TX [33] pkt 0x20407290 ref 2 (eth_tx():1417)
[00:00:03.827,124] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TX [34] pkt 0x20407290 ref 0 frags 0x20408618 (ethernet_send():536)
[00:00:03.838,962] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [32] frag 0x20408618
[00:00:03.848,115] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref -1 frags 0x00000000 (ethernet_send():536)
[00:00:03.858,230] [1;31m<err> net_pkt.net_pkt_unref_debug: *** ERROR *** frag 0x20408618 is freed already by tx_completed():554 (ethernet_send():536)[0m
[00:00:03.871,859] <dbg> net_pkt.net_pkt_frag_unref_debug: (0x20401c80): TDATA (tx_bufs) [32] frag 0x20408618 ref -1 (net_pkt_unref_debug():823)
[9[[00:0:00:03.884,23685] <;dbg> net_pkt.net_pkt_get_reser_ve_debug:  (0x20402e74):g TX2 [034]  pskfte 0x20407290 r ex_soeprevee( 14 r eefh 1 (net_ipv6_sen)d_rs():2119)
r[00:00:03.899,402] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74): TDATA (tx_bufs) [31] frag 0x20408600 reserve 14
[00:00:03.910,497] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20402e74):   ref 1 (net_ipv6_send_rs():2124)
[00:00:03.920,344] <dbg> net_pkt.net_pkt_frag_add_debug: (0x20402e74): pkt 0x20407290 frag 0x20408600 (net_ipv6_send_rs:2130)
[00:00:03.931,403] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618 reserve 14
[00:00:03.942,379] <dbg> net_pkt.net_pkt_get_reserve_data_debug: (0x20401c80):   ref 1 (ethernet_fill_header():408)
[00:00:03.952,491] <dbg> net_pkt.net_pkt_frag_insert_debug: (0x20401c80): pkt 0x20407290 frag 0x20408618 (ethernet_fill_header:462)
[00:00:03.964,008] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TX [34] pkt 0x20407290 ref 0 frags 0x20408618 (tx_completed():554)
[00:00:03.965,733] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618
[00:00:03.964,887] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref 0 frags 0x20408600 (tx_completed():554)
[00:00:03.964,830] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408600
[00:00:03.963,980] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref 0 frags 0x00000000 (tx_completed():554)
[00:00:03.963,931] <dbg> net_pkt.net_pkt_frag_unref_debug: (0x20401c80): TDATA (tx_bufs) [30] frag 0x20408618 ref 0 (net_pkt_unref_debug():823)
[00:00:03.976,497] <dbg> net_pkt.net_pkt_ref_debug: (0x20401c80): TX [35] pkt 0x20407290 ref 1 (eth_tx():1417)
[00:00:03.986,126] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TX [35] pkt 0x20407290 ref 0 frags 0x20408618 (ethernet_send():536)
[00
:00:03.997,963] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80): TDATA (tx_bufs) [32] frag 0x20408618
[00:00:04.007,116] <dbg> net_pkt.net_pkt_unref_debug: (0x20401c80):   ref -1 frags 0x00000000 (ethernet_send():536)
[00:00:04.017,232] [1;31m<err> net_pkt.net_pkt_unref_debug: *** ERROR *** frag 0x20408618 is freed already by tx_completed():554 (ethernet_send():536)[0m
[00:00:04.030,861] <dbg> net_pkt.net_pkt_frag_unref_debug: (0x20401c80): TDATA (tx_bufs) [32] frag 0x20408618 ref -1 (net_pkt_unref_debug():823)
[9D[J[00:00:04.044,008] <dbg> net_pkt.net_pkt_unref_debug: (0x204007b0): RX [31] pkt 0x20406490 ref 0 frags 0x20407300 (gptp_thread():554)
[00:00:04.055,678] <dbg> net_pkt.net_pkt_unref_debug: (0x204007b0): RDATA (rx_bufs) [13] frag 0x20407300
[00:00:04.064,832] <dbg> net_pkt.net_pkt_unref_debug: (0x204007b0):   ref 0 frags 0x00000000 (gptp_thread():554)
[00:00:04.074,692] <dbg> net_pkt.net_pkt_frag_unref_debug: (0x204007b0): RDATA (rx_bufs) [13] frag 0x20407300 ref 0 (net_pkt_unref_debug():823)
```

Commit:
net/net_if: Do not set TCP related flags on non-INET/6 family

Such sent flag is in a union in net_pkt, shared with a gptp flag.
Tweaking it when the family is not AF_INET or AF_INET6 will generate
corrupted gptp packets.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>
Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar reviewed (approved):
LGTM, also the latest version fixes memory leak in sam-e70 board

pfalcon commented on subsys/net/l2/ethernet/arp.c:
```diff
@@ -252,33 +251,28 @@ static inline struct net_pkt *arp_prepare(struct net_if *iface,
 
-	eth = net_eth_fill_header(ctx, pkt, htons(NET_ETH_PTYPE_ARP),
-				  NULL, NULL);
+	if(!net_eth_fill_header(ctx, pkt, htons(NET_ETH_PTYPE_ARP),
```
So, it went to the mainline like that, with missing space after "if"?

pfalcon commented on include/net/net_l2.h:
```diff
@@ -49,8 +49,9 @@ struct net_l2 {
 	 * This function is used by net core to push a packet to lower layer
 	 * (interface's L2), which in turn might work on the packet relevantly.
 	 * (adding proper header etc...)
+	 * Returns a negative error code, or the number of bytes sent otherwise.
```
Driver patches as included in this commit don't seem to follow "the number of bytes sent otherwise", so stats would be off (unless I'm mixing up all the send()'s around).

pfalcon commented on subsys/net/l2/ieee802154/ieee802154.c:
```diff
@@ -244,42 +245,57 @@ static enum net_verdict ieee802154_recv(struct net_if *iface,
 
 	if (net_pkt_family(pkt) != AF_INET6) {
-		return NET_DROP;
+		return -EINVAL;
```
Maybe -EPROTONOTSUP or how it's called?

pfalcon commented on include/net/net_pkt.h:
```diff
@@ -118,10 +118,18 @@ struct net_pkt {
-				 * to be sent but has not reached the driver
-				 * yet. Used only if defined(CONFIG_NET_TCP)
-				 */
+	union {
```
Are you sure bitfield union works as expected? I'd expect compiler bugs.

pfalcon commented:
@tbursztyka, I'm curious how well this was tested. Changes to existing Ethernet drivers are pretty light, whereas this refactor immediately breaks my cute WIP driver https://github.com/zephyrproject-rtos/zephyr/pull/11680, where I took freedom to rely on the fact that any net_buf (except last) has len of even u32_t's. Apparently, any optimization in Zephyr is premature so far ;-). But I wonder if no other driver had anything like that?

tbursztyka commented:
@pfalcon would have been nice to get your review before it got merged. This PR has been public for more than a month.

pfalcon commented:
> This PR has been public for more than a month.

Really, in my list it's up since ~ end of June, before you went on leave, though maybe it was just a branch, not PR then. Who'd have known that would be merged last week? ;-)

Otherwise, I kinda skip such reviews, because you definitely already heard my comments: this is too complex, can we split it up, etc. I just accept that some changes are necessarily complex, and rely on our discussion in Le Mans that it's otherwise OK. You're absolutely right that I should at least eyeball them for simple issues nonetheless.
